osf_tests/utils.py

```diff
@@ -176,3 +177,21 @@ def __init__(self, status_code):
     def raise_for_status(self):
         if self.status_code >= 400:
             raise Exception
+
+def create_mock_blob():
+    mock_blob = mock.create_autospec(Blob)
+    mock_blob.delete.return_value = None
+    return mock_blob
+
+def create_mock_bucket():
+    mock_bucket = mock.create_autospec(Bucket)
+    mock_bucket.get_blob.return_value = create_mock_blob()
+    return mock_bucket
+
+def create_mock_gcs_client():
```

Comment:
If possible mock the client using `responses` instead of creating an auto-spec mock. I just makes it more explicit what payload are being sent where and it prevents over-mocking something so it passes every time.

Reply:
It's true that the tests assume 1) Google's GCS client library _works_ and 2) the data model referencing cloud storage location is consistent, but the point of using MagicMocks for the GCS client is to avoid testing GCS client internals, focusing more on confirming correctness of the application logic that determines whether or not to actually purge something. Namely, that shared or duplicated FileVersion objects attached to live filenodes are avoided.

Reply:
IMO the strongest likelihood that this will break probably comes from the fact that the client is a black box and yet we will probably have to update it a few times to keeps this working over the years, but obviously that's a good point as it doesn't really matter in the scope of this issue.

Reply:
`create_autospec` ensures the consistency of function existence and parameters, at least:
```
In [1]: from google.cloud.storage import Client, Bucket, Blob

In [2]: import mock

In [3]: cli = mock.create_autospec(Client)

In [4]: cli.not_a_real_fn.return_value = mock.create_autospec(Bucket)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-f26054499a65> in <module>
----> 1 cli.not_a_real_fn.return_value = mock.create_autospec(Bucket)

~/.pyenv/versions/3.6.2/envs/osf/lib/python3.6/site-packages/mock/mock.py in __getattr__(self, name)
    696         elif self._mock_methods is not None:
    697             if name not in self._mock_methods or name in _all_magics:
--> 698                 raise AttributeError("Mock object has no attribute %r" % name)
    699         elif _is_magic(name):
    700             raise AttributeError(name)

AttributeError: Mock object has no attribute 'not_a_real_fn'

In [5]: cli.get_bucket.return_value = mock.create_autospec(Bucket)

In [6]: cli.get_bucket()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-e1453bbdfc9d> in <module>
----> 1 cli.get_bucket()

~/.pyenv/versions/3.6.2/envs/osf/lib/python3.6/site-packages/mock/mock.py in __call__(_mock_self, *args, **kwargs)
   1059         # can't use self in-case a function / method we are mocking uses self
   1060         # in the signature
-> 1061         _mock_self._mock_check_sig(*args, **kwargs)
   1062         return _mock_self._mock_call(*args, **kwargs)
   1063

~/.pyenv/versions/3.6.2/envs/osf/lib/python3.6/site-packages/mock/mock.py in checksig(_mock_self, *args, **kwargs)
    207     func, sig = sig
    208     def checksig(_mock_self, *args, **kwargs):
--> 209         sig.bind(*args, **kwargs)
    210     _copy_func_details(func, checksig)
    211     type(mock)._mock_check_sig = checksig

~/.pyenv/versions/3.6.2/lib/python3.6/inspect.py in bind(*args, **kwargs)
   2963         if the passed arguments can not be bound.
   2964         """
-> 2965         return args[0]._bind(args[1:], kwargs)
   2966
   2967     def bind_partial(*args, **kwargs):

~/.pyenv/versions/3.6.2/lib/python3.6/inspect.py in _bind(self, args, kwargs, partial)
   2878                             msg = 'missing a required argument: {arg!r}'
   2879                             msg = msg.format(arg=param.name)
-> 2880                             raise TypeError(msg) from None
   2881             else:
   2882                 # We have a positional argument to process

TypeError: missing a required argument: 'bucket_name'
```